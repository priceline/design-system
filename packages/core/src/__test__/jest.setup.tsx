import '@testing-library/jest-dom'
import 'jest-styled-components'

global.requestAnimationFrame =
  global.requestAnimationFrame ||
  function _raf(cb) {
    return setTimeout(cb, 0)
  }
