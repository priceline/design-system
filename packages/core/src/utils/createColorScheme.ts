import { ColorSchemes, PaletteFamilies } from '../theme'

export const createColorScheme = ({ palette }: { palette: PaletteFamilies }): ColorSchemes => {
  const {
    primary,
    secondary,
    background,
    success,
    warning,
    caution,
    text,
    highlight,
    promoPrimary,
    promoSecondary,
    alert,
  } = palette

  return {
    // Primary
    ///////////////////////////////////////////////
    primary: {
      background: primary.base,
      backgroundName: 'primary.base',
      foreground: background.lightest,
    },
    primaryLight: {
      background: primary.light,
      backgroundName: 'primary.light',
      foreground: primary.base,
    },
    primaryLightest: {
      background: background.lightest,
      backgroundName: 'background.lightest',
      foreground: primary.base,
    },
    primaryDark: {
      background: primary.dark,
      backgroundName: 'primary.dark',
      foreground: background.lightest,
    },
    primaryShade: {
      background: primary.shade,
      backgroundName: 'primary.shade',
      foreground: background.lightest,
    },
    primaryDarkOnLight: {
      background: primary.light,
      backgroundName: 'primary.light',
      foreground: primary.dark,
    },

    // Secondary
    ///////////////////////////////////////////////
    secondary: {
      background: secondary.base,
      backgroundName: 'secondary.base',
      foreground: background.lightest,
    },
    secondaryLight: {
      background: secondary.light,
      backgroundName: 'secondary.light',
      foreground: secondary.base,
    },
    secondaryLightest: {
      background: background.lightest,
      backgroundName: 'background.lightest',
      foreground: secondary.base,
    },
    secondaryDark: {
      background: secondary.dark,
      backgroundName: 'secondary.dark',
      foreground: background.lightest,
    },
    secondaryDarkOnLight: {
      background: secondary.light,
      backgroundName: 'secondary.light',
      foreground: secondary.dark,
    },

    // Neutral
    ///////////////////////////////////////////////
    neutral: {
      background: background.dark,
      backgroundName: 'background.dark',
      foreground: background.lightest,
    },
    neutralLight: {
      background: background.base,
      backgroundName: 'background.base',
      foreground: background.dark,
    },
    neutralLightest: {
      background: background.lightest,
      backgroundName: 'background.lightest',
      foreground: background.dark,
    },
    neutralDark: {
      background: background.darkest,
      backgroundName: 'background.darkest',
      foreground: background.lightest,
    },
    neutralDarkOnLight: {
      background: background.light,
      backgroundName: 'background.light',
      foreground: background.darkest,
    },
    neutralDarkOnLightest: {
      background: background.lightest,
      backgroundName: 'background.lightest',
      foreground: background.darkest,
    },

    // Success
    ///////////////////////////////////////////////
    success: {
      background: success.base,
      backgroundName: 'success.base',
      foreground: background.lightest,
    },
    successLight: {
      background: success.light,
      backgroundName: 'success.light',
      foreground: success.base,
    },
    successLightest: {
      background: background.lightest,
      backgroundName: 'background.lightest',
      foreground: success.base,
    },
    successDark: {
      background: success.dark,
      backgroundName: 'success.dark',
      foreground: background.lightest,
    },
    successDarkOnLight: {
      background: success.light,
      backgroundName: 'success.light',
      foreground: success.dark,
    },

    // Warning
    ///////////////////////////////////////////////
    warning: {
      background: warning.base,
      backgroundName: 'warning.base',
      foreground: background.lightest,
    },
    warningLight: {
      background: warning.light,
      backgroundName: 'warning.light',
      foreground: warning.base,
    },
    warningLightest: {
      background: background.lightest,
      backgroundName: 'background.lightest',
      foreground: warning.base,
    },

    // Caution
    //////////////////////////////////////////////
    cautionLight: {
      background: caution.light,
      backgroundName: 'caution.light',
      foreground: text.base,
    },

    // Highlight
    //////////////////////////////////////////////
    highlightLight: {
      background: highlight.light,
      backgroundName: 'highlight.light',
      foreground: secondary.dark,
    },

    // Promo
    //////////////////////////////////////////////
    promo: {
      background: promoPrimary.base,
      backgroundName: 'promoPrimary.base',
      foreground: text.lightest,
    },
    promoLight: {
      background: promoPrimary.light,
      backgroundName: 'promoPrimary.light',
      foreground: promoPrimary.base,
    },
    promoLightest: {
      background: background.lightest,
      backgroundName: 'background.lightest',
      foreground: promoPrimary.base,
    },
    promoDark: {
      background: promoPrimary.dark,
      backgroundName: 'promoPrimary.dark',
      foreground: text.lightest,
    },
    promoDarkOnLight: {
      background: promoPrimary.light,
      backgroundName: 'promoPrimary.light',
      foreground: promoPrimary.dark,
    },

    // Promo Secondary
    //////////////////////////////////////////////
    promoSecondary: {
      background: promoSecondary.base,
      backgroundName: 'promoSecondary.base',
      foreground: text.lightest,
    },
    promoSecondaryDark: {
      background: promoSecondary.dark,
      backgroundName: 'promoSecondary.dark',
      foreground: text.lightest,
    },

    // Alert
    //////////////////////////////////////////////
    alert: {
      background: alert.base,
      backgroundName: 'alert.base',
      foreground: background.lightest,
    },
    alertLight: {
      background: alert.light,
      backgroundName: 'alert.light',
      foreground: alert.base,
    },
    alertLightest: {
      background: background.lightest,
      backgroundName: 'background.lightest',
      foreground: alert.base,
    },
    alertDarkOnLight: {
      background: alert.light,
      backgroundName: 'alert.light',
      foreground: alert.tone,
    },

    // Heading
    //////////////////////////////////////////////
    heading: {
      background: background.lightest,
      backgroundName: 'background.lightest',
      foreground: text.heading,
    },
    headingBase: {
      background: background.base,
      backgroundName: 'background.base',
      foreground: text.heading,
    },
    headingLight: {
      background: background.light,
      backgroundName: 'background.light',
      foreground: text.heading,
    },
    headingOnPrimaryLight: {
      background: primary.light,
      backgroundName: 'primary.light',
      foreground: text.heading,
    },
  }
}
