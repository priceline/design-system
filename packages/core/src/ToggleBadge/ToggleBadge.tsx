import React, { type ComponentPropsWithRef } from 'react'
import styled from 'styled-components'
import { FontSizeProps, SpaceProps, borderRadius, compose, fontSize, space } from 'styled-system'
import { borderRadiusAttrs } from '../utils/attrs'
import { applyVariations, getPaletteColor } from '../utils/utils'

/**
 * @public
 */
export type ToggleBadgeProps = {
  borderRadius?: string
  selected?: boolean
  color?: string
  bg?: string
  unSelectedBg?: string
} & SpaceProps &
  FontSizeProps &
  ComponentPropsWithRef<'button'>

/**
 * @public
 */
export const ToggleBadge: React.FC<ToggleBadgeProps> = styled.button.attrs(borderRadiusAttrs)`
  border-radius: ${(props) => props.theme.radius};
  border: 0;
  display: inline-block;
  font-weight: ${(props) => props.theme.bold};
  font-family: inherit;
  cursor: pointer;
  background-color: ${(props) =>
    props.selected ? getPaletteColor(props.bg || props.color, 'light')(props) : props.unSelectedBg};
  color: ${getPaletteColor('base')};
  &:hover {
    background-color: ${(props) => getPaletteColor(props.bg || props.color, 'light')(props)};
  }
  ${applyVariations('ToggleBadge')}

  ${(props) => compose(space, fontSize, borderRadius)(props)}
`

ToggleBadge.displayName = 'ToggleBadge'

ToggleBadge.defaultProps = {
  borderRadius: 'full',
  selected: false,
  px: 2,
  py: 1,
  mx: 1,
  my: 1,
  fontSize: 0,
  color: 'primary',
  unSelectedBg: 'transparent',
}
