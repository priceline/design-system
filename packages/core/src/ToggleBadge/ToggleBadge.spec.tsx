import { render, screen } from '../__test__/testing-library'
import { ToggleBadge } from '../ToggleBadge/ToggleBadge'
import { createTheme } from '../utils/createTheme'

describe('ToggleBadge', () => {
  let consoleError
  let theme
  beforeEach(() => {
    consoleError = console.error
    console.error = jest.fn()
    theme = createTheme()
  })
  afterEach(() => (console.error = consoleError))

  test('selected ToggleBadge renders with default props', () => {
    render(<ToggleBadge selected data-testid='toggle-badge' />)
    const element = screen.getByTestId('toggle-badge')
    expect(element).toHaveStyleRule('background-color', theme.colors.lightBlue)
    expect(element).toHaveStyleRule('color', theme.colors.blue)
  })

  test('unselected ToggleBadge renders with default props', () => {
    render(<ToggleBadge data-testid='toggle-badge' />)
    const element = screen.getByTestId('toggle-badge')
    expect(element).toHaveStyleRule('color', theme.colors.blue)
  })

  test('selected one with background-color and text color passed as props hover state', () => {
    render(<ToggleBadge selected bg='secondary' color='error' fontSize={1} data-testid='toggle-badge' />)
    const element = screen.getByTestId('toggle-badge')
    expect(element).toHaveStyleRule('background-color', theme.palette.secondary.light)
    expect(element).toHaveStyleRule('color', theme.palette.error.base)
  })
})
