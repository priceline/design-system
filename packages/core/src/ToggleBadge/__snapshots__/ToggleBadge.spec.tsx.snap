// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleBadge selected ToggleBadge renders with default props 1`] = `
.c0 {
  border-radius: 2px;
  border: 0;
  display: inline-block;
  font-weight: 700;
  font-family: inherit;
  cursor: pointer;
  background-color: #e8f2ff;
  color: #0068ef;
  border-radius: 9999px;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 4px;
  padding-bottom: 4px;
  margin-left: 4px;
  margin-right: 4px;
  margin-top: 4px;
  margin-bottom: 4px;
  font-size: 12px;
}

.c0:hover {
  background-color: #e8f2ff;
}

<button
  className="c0"
  color="primary"
  fontSize={0}
  selected={true}
/>
`;

exports[`ToggleBadge selected one with background-color and text color passed as props hover state 1`] = `
.c0 {
  border-radius: 2px;
  border: 0;
  display: inline-block;
  font-weight: 700;
  font-family: inherit;
  cursor: pointer;
  background-color: #ecf7ec;
  color: #c00;
  font-size: 14px;
  border-radius: 9999px;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 4px;
  padding-bottom: 4px;
  margin-left: 4px;
  margin-right: 4px;
  margin-top: 4px;
  margin-bottom: 4px;
}

.c0:hover {
  background-color: #ecf7ec;
}

<button
  className="c0"
  color="error"
  fontSize={1}
  selected={true}
/>
`;

exports[`ToggleBadge unselected ToggleBadge renders with default props 1`] = `
.c0 {
  border-radius: 2px;
  border: 0;
  display: inline-block;
  font-weight: 700;
  font-family: inherit;
  cursor: pointer;
  background-color: transparent;
  color: #0068ef;
  border-radius: 9999px;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 4px;
  padding-bottom: 4px;
  margin-left: 4px;
  margin-right: 4px;
  margin-top: 4px;
  margin-bottom: 4px;
  font-size: 12px;
}

.c0:hover {
  background-color: #e8f2ff;
}

<button
  className="c0"
  color="primary"
  fontSize={0}
  selected={false}
/>
`;
